prost-build/build.rs

```diff
@@ -1,137 +1,35 @@
-extern crate curl;
-extern crate tempdir;
-extern crate zip;
+extern crate fs_extra;
+
+use fs_extra::dir;
 
 use std::env;
 use std::fs;
-use std::io::{
-    self,
-    Cursor,
-};
 use std::path::{
-    Path,
     PathBuf,
 };
 
-use curl::easy::Easy;
-use tempdir::TempDir;
-use zip::ZipArchive;
-
-const VERSION: &'static str = "3.5.1";
-
 fn main() {
-    if env_contains_protoc() { return; }
-
-    let out_dir = PathBuf::from(env::var("OUT_DIR").expect("OUT_DIR environment variable is invalid"));
-    let protoc_dir = out_dir.join("protoc");
-
-    // If the protoc directory doesn't already exist from a previous build, download protoc.
-    if !protoc_dir.exists() {
-        // The unzipping done by download_protoc is not atomic, so to avoid failing halfway through
-        // and leaving trash state, download/unzip to a temporary directory, then rename.
-        let tempdir = TempDir::new_in(&out_dir, "protoc").expect("failed to create temporary directory");
-        download_protoc(tempdir.path());
-        fs::rename(&tempdir.into_path(), &protoc_dir).expect("unable to move temporary directory");
-    }
-    let mut protoc = protoc_dir.join("bin");
-    protoc.push("protoc");
-    protoc.set_extension(env::consts::EXE_EXTENSION);
-
-    println!("cargo:rustc-env=PROTOC={}", protoc.display());
-    println!("cargo:rustc-env=PROTOC_INCLUDE={}", protoc_dir.join("include").display());
-    println!("cargo:rerun-if-env-changed=PROTOC");
-    println!("cargo:rerun-if-env-changed=PROTOC_INCLUDE");
-}
-
-/// Returns `true` if the environment already contains the `PROTOC` and `PROTOC_INCLUDE` variables.
-fn env_contains_protoc() -> bool {
-    let protoc = match env::var("PROTOC") {
-        Ok(val) => PathBuf::from(val),
-        Err(env::VarError::NotPresent) => return false,
-        Err(env::VarError::NotUnicode(..)) => panic!("PROTOC environment variable is not valid UTF-8"),
-    };
-
-    let protoc_include = match env::var("PROTOC_INCLUDE") {
-        Ok(val) => PathBuf::from(val),
-        Err(env::VarError::NotPresent) => panic!("PROTOC_INCLUDE environment variable not set (PROTOC is set)"),
-        Err(env::VarError::NotUnicode(..)) => panic!("PROTOC_INCLUDE environment variable is not valid UTF-8"),
+    // Recursively copy all the protobuf files to `OUT_DIR` so that they are
+    // available for use at runtime. The binaries for all platforms are copied
```

Comment:
> The binaries for all platforms...

I don't really follow this reasoning, maybe clarify it a bit?

I think an alternate strategy could be to create a crate per platform that each bundle a single `protoc` and have Cargo include the right one through platform `cfg`s, but honestly that's going to complicate the publishing and updating process so much that I'd rather keep it as is.

Reply:
Sure. In my case I use Windows 10 and so I have Win32 and Microsoft Linux host architectures that use the same (target) output directories on the same system. In another case I have Linux using SAMBA to share the directories with a remote Windows system. I'm trying to find some reasonable way to support this kind of situation.